zkevm-circuits/src/root_circuit/aggregation.rs

```diff
@@ -0,0 +1,698 @@
+        self.proof
+    }
+
+    fn loaded_instances<'b>(
+        &self,
+        loader: &Rc<Halo2Loader<'b, C>>,
+    ) -> Vec<Vec<Scalar<'b, C, EccChip<C>>>> {
+        self.instances
+            .transpose_vec(self.protocol.num_instance.len())
+            .into_iter()
+            .zip(self.protocol.num_instance.iter())
+            .map(|(instances, num_instance)| {
+                instances
+                    .transpose_vec(*num_instance)
+                    .into_iter()
+                    .map(|instance| loader.assign_scalar(instance.copied()))
+                    .collect_vec()
+            })
+            .collect_vec()
+    }
+}
+
+/// Aggregation configuration.
+#[derive(Clone)]
+pub struct AggregationConfig {
+    /// MainGateConfig
+    pub main_gate_config: MainGateConfig,
+    /// RangeConfig
+    pub range_config: RangeConfig,
+}
+
+impl AggregationConfig {
+    /// Configure for `MainGate` and `RangeChip` with corresponding fixed lookup
+    /// table.
+    pub fn configure<C: CurveAffine>(meta: &mut ConstraintSystem<C::Scalar>) -> Self {
+        let main_gate_config = MainGate::<C::Scalar>::configure(meta);
+        let range_config = RangeChip::<C::Scalar>::configure(
+            meta,
+            &main_gate_config,
+            vec![BITS / LIMBS],
+            Rns::<C::Base, C::Scalar, LIMBS, BITS>::construct().overflow_lengths(),
+        );
+        Self {
+            main_gate_config,
+            range_config,
+        }
+    }
+
+    /// Returns `MainGate`.
+    pub fn main_gate<F: FieldExt>(&self) -> MainGate<F> {
+        MainGate::new(self.main_gate_config.clone())
+    }
+
+    /// Returns `RangeChip`.
+    pub fn range_chip<F: FieldExt>(&self) -> RangeChip<F> {
+        RangeChip::new(self.range_config.clone())
+    }
+
+    /// Returns `EccChip`.
+    pub fn ecc_chip<C: CurveAffine>(&self) -> EccChip<C> {
+        EccChip::new(EccConfig::new(
+            self.range_config.clone(),
+            self.main_gate_config.clone(),
+        ))
+    }
+
+    /// Load fixed lookup table for `RangeChip`
+    pub fn load_table<F: FieldExt>(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        self.range_chip().load_table(layouter)
+    }
+
+    /// Aggregate snarks into a single accumulator and decompose it into
+    /// `4 * LIMBS` limbs.
+    /// Returns assigned instances of snarks and aggregated accumulator limbs.
+    #[allow(clippy::type_complexity)]
+    pub fn aggregate<'a, M: MultiMillerLoop>(
+        &self,
+        layouter: &mut impl Layouter<M::Scalar>,
+        svk: &KzgSvk<M>,
+        snarks: impl IntoIterator<Item = SnarkWitness<'a, M::G1Affine>>,
+    ) -> Result<
+        (
+            Vec<Vec<Vec<AssignedCell<M::Scalar, M::Scalar>>>>,
+            Vec<AssignedCell<M::Scalar, M::Scalar>>,
+        ),
+        Error,
+    > {
+        type PoseidonTranscript<'a, C, S> =
+            transcript::halo2::PoseidonTranscript<C, Rc<Halo2Loader<'a, C>>, S, T, RATE, R_F, R_P>;
```

Comment:
I think for `zkevm-circuits` the only place using this is here, so I didn't export it, the main reason is to allow exporting `PoseidonTranscript` with `NativeLoader` for easier `create_proof ` and `verify_proof` usage, otherwise people need to specify `NativeLoader` manually which seems non-sense and annoying imo.

Reply:
Ohhh makes sense! Thanks!